### ISSUE TITLE
Datastore not saving @hasOne model - 'data couldn't be read because it is missing'

### ISSUE THREAD

**Issue by dsakhapo** [aws-amplify/amplify-swift#2436]
### Describe the bug

I set up a basic schema:

```
type Message @model {
  id: ID!
  message: String!
  from: User! @hasOne
  channelID: ID!
}

type User @model {
  id: ID!
  userID: ID!
  email: String!
}
```

I am using the Datastore to try and store a Message which has a User and I am receiving the following output in the console:

> DataStoreError: The data couldn’t be read because it is missing.
Recovery suggestion: The data couldn’t be read because it is missing.
Caused by:
keyNotFound(CodingKeys(stringValue: "from", intValue: nil), Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "elements", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0)], debugDescription: "No value associated with key CodingKeys(stringValue: \"from\", intValue: nil) (\"from\").", underlyingError: nil))

When I make the User optional in the Message type, I do not receive this error but the value for User will always be nil despite checking the Message object prior to Datastore.save().

I have noticed after using 

> amplify codegen models

the Message type will add a type `messageFromId: ID` and I have tried multiple ways to create the Message object:

```
func createMessage(channelID: String, message: String, from: User) {
            let storeMessage = Message(message: message, messageFromId: from.id, from: from, channelID: channelID)
            print("createMessage(): \(storeMessage)")

            let _ = Amplify.DataStore.save(storeMessage)
            .sink {
                if case let .failure(error) = $0 {
                    print("Error saving message \(error)")
                }
            } receiveValue: { message in
                print("Message saved successfully: \(message)")
            }
}
```

Above where I create the `storeMessage` object I have tried omitting the `messageFromId` from creation while still passing in the User, and also vise-versa (when I tried making the User optional).

I have also tried enabling AppSync and updating the api via:

> amplify push

which will complete successfully but these issues still remain.

Any assistance with this issue is greatly appreciated - thank you.

### Steps To Reproduce

```swift
Use this schema and try to save a Message in the Datastore with a @hasOne User.
```

### Expected behavior

The datastore will save models with their related types, and can be queried.

### Amplify Framework Version

1.28.1

### Amplify Categories

DataStore

### Dependency manager

Swift PM

### Swift version

5.6.1

### CLI version

10.2.3

### Xcode version

13.4.1 (13F100)

### Relevant log output

```shell
<details>
<summary>Log Messages</summary>


INSERT LOG MESSAGES HERE
```
</details>
```


### Is this a regression?

No

### Regression additional context

_No response_

### Device

iPhone 13 Pro - Simulator

### iOS Version

15.4

### Specific to simulators

_No response_

### Additional context

_No response_

**Comment by lawmicha** [comment]
Hi @dsakhapo,

Sorry for the confusion on how to save the Message type. Currently the codegen process exposes the foreign key `messageFromId` for hasOne relationships, while it doesn't do this other types of associations like hasMany/belongs-to. We may want to change this to align the codegenerated types with hasOne to the latter so that there's only one place to create the hasOne relationship. Currently, it's confusing given the Message initializer, do we pass in the User object for the `from` parameter, the User's identifier to the `messageFromId` parameter, or both? There's an opportunity here to document how to achieve the CRUD operations with DataStore given this model.

What I observed[1](https://github.com/aws-amplify/amplify-ios/blob/dev-preview/AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginV2Tests/TransformerV2/DataStoreConnectionScenario1V2Tests.swift#L108-L114) is that you can just pass in the identifier.
```
let message = Message(message: message, messageFromId: from.id, from: nil, channelID: channelID)
Amplify.DataStore.save(message)
```
When you query for the Message, the `messageFromId` will be populated, which you can then query for the User with [2](https://github.com/aws-amplify/amplify-ios/blob/dev-preview/AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginV2Tests/TransformerV2/DataStoreConnectionScenario1V2Tests.swift#L150)
```
let message = /* Amplify.DataStore.query byid, or list and access one */
let user = Amplify.DataStore.query(User.self, byId: message.messageFromid)
```

Can you confirm if this works for you?

**Comment by dsakhapo** [comment]
Hello @lawmicha 

Thank you for your reply. To answer your first question, I tried multiple ways of creating a Message object because I wasn't sure of the behavior:

> Method 1 - Only pass in the `messageFromId`

`let message = Message(message: message, messageFromId: from.id, channelID: channelID)`

I tried this method because I wasn't sure whether or not the associated User would be added to the Message table.

> Method 2 - Only pass in the User i.e. `from`

`let message = Message(message: message, from: from, channelID: channelID)`

Similar to above, I wasn't sure if the messageFromId field would be added by Amplify using the User's primary key.

> Method 3 - pass in both the User and foreign key i.e. `from` and `messageFromId`

`let message = Message(message: message, messageFromId: from.id, from: from, channelID: channelID)`

Neither of these approaches worked for me, as the `from` field in a Message would always be `nil` in the event that User was optional, or I received the error from the console in the post if User was required.

To answer your second question, passing in the identifier `messageFromId` and then querying the User table using this Id does work, however, I was wondering if it was possible to query a Message and access the User tied to it directly instead of doing a separate query?

In other words, if I query a Message, can I access the associated User directly or do I need two separate queries if I want the relationship data? Also, I'd like to make the User required if possible.

Thanks for your assistance.

**Comment by aspearman** [comment]
Hi, I had the exact same confusion. 

```
type Event @model {
  id: ID!
  recipe: Recipe! @hasOne
  ...
}

type Recipe @model {
  id: ID!
  name: String!
  ...
}
```

To construct an Event, I do:
```
let event = Event(
                recipe: recipe,
                ...
                eventRecipeId: recipe.id
            )
```

However I still receive the error immediately on `Amplify.DataStore.save`:
```
Failed to save event with error: DataStoreError: The data couldn’t be read because it is missing.
Recovery suggestion: The data couldn’t be read because it is missing.
Caused by:
keyNotFound(CodingKeys(stringValue: "recipe", intValue: nil), Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "elements", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0)], debugDescription: "No value associated with key CodingKeys(stringValue: \"recipe\", intValue: nil) (\"recipe\").", underlyingError: nil))
```

It'd definitely be nice if I didn't have to manually pass in the `eventRecipeId` like you said @lawmicha, or at least see some documentation about what to pass in. 

Nonetheless, passing in `recipe.id` like you suggested still results in an error, so I'm not sure what to do. It's upset that my `recipe` is `nil` which if I `po recipe` just before saving shows that's not true; it has values. I'm not sure of the internals of what happens when it saves (maybe I'll read through the GitHub repo...) so maybe it queries to try to find the `recipe` but it can't for some reason? Even though I'm providing `recipe.id` as its relationship ID, so it should be able to find the recipe. I'm not sure why this doesn't work. 

@dsakhapo For me there's only one constructor, which requires both fields, so I'm not sure how you managed to try passing in just one of either `messageFromId` or `from`. You said on Discord that the "workaround" worked? It's not working for me to pass in `recipe.id`...

**Comment by dsakhapo** [comment]
Hi @aspearman 

The workaround only works if you make the `recipe` optional:

```
type Event @model {
  id: ID!
  recipe: Recipe @hasOne
  ...
}

type Recipe @model {
  id: ID!
  name: String!
  ...
}
```

Which you would then pass in the `recipe.id` for the `eventRecipeId` field, while passing in `nil` for the `recipe`:

`let event = Event(recipe: nil, eventRecipeId: recipe.id)` 

which you could then query for a `Recipe` using the `eventRecipeId`:

`let recipe = Amplify.Datastore.Query(Recipe.self, byId: event.eventRecipeId)`

That being said though, I would like to know if it is possible to query for an `Event` and retrieve the associated `Recipe`. Because in this workaround it seems you wouldn't even need to use the `@hasOne` directive and instead could just create a field to store the foreign key of whatever type you would like to have a relationship to.

Another question is whether or not we can make relationships required, as we are needing to make it optional for this to work.

@lawmicha Any clarifications on these questions would be greatly appreciated! Thank you.

**Comment by aspearman** [comment]
Thanks @dsakhapo. I've worked around it using a `@belongsTo` and `@hasMany` relationship instead, because I want to make my relationship non-optional, and I don't really want to do a separate query for the `recipe` property. 

Seems like this is a bug with `@hasOne`, though an Amplify dev on Discord mentioned `@hasOne` for some reason has a different initializer than the others i.e. you must pass in both the recipe and recipe.id as relationship id, unlike other relationship types like `@belongsTo` where it just needs `recipe`, so perhaps it's just an inconsistency. Would be really nice to make `@hasOne` behave more like the other relationships.

**Comment by aspearman** [comment]
@phil-margetson If it helps, I ended up going with a bi-directional `@belongsTo`.

**Comment by bjkillen** [comment]
Would also like an update on this issue. Just migrated from V1 today and sad to see the @hasMany works different than the connection direction implementation. We have been building tons of code around the assumption the relationships look the same from the outside so frustrating to have to burn time doing workarounds like this just to get back to a previous behavior.

**Comment by thisisabhash** [comment]
Hello, we apologize for the delay in our response and thank you for your patience in this. 

We have made several improvements to Datastore and API in recent releases [specially in v2.4.0, where we release Lazy Loading support for connected models]. Please follow the steps here to update your CLI and Amplify version to make use of this feature. 
https://docs.amplify.aws/cli/migration/lazy-load-custom-selection-set/#lazy-loading-connected-models

I tried to reproduce the problem using the schema you mentioned above and could no longer reproduce it.  For reference, I used the code snippet below and attaching a screenshot from debugger [all the saved and queried items are populated successfully] : 

```
func createAndQueryMessage() async {
    let user = User(userID: "user", email: "user@user.com")
    let message = Message(message: "message from user",
                          from: user,
                          channelID: "channelID",
                          messageFromId: user.id)
    do {
        let savedUser = try await Amplify.DataStore.save(user)
        print("Saved user: \(savedUser)")
        let savedMessage = try await Amplify.DataStore.save(message)
        print("Saved message: \(savedMessage)")
        
        let queriedMessageOptional = try await Amplify.DataStore.query(Message.self, byId: savedMessage.id)
        guard let queriedMessage = queriedMessageOptional else {
            print("Queried message is nil")
            return
        }
        print("Queried message: \(queriedMessage)")
        
        let userFromQueriedMessage = try await queriedMessage.from
        print("User from queried message: \(userFromQueriedMessage)")
    } catch (let error) {
        print("Failed to perform operation with: \(error)")
    }
}
```
<img width="1242" alt="screenshot-hasone" src="https://user-images.githubusercontent.com/23294582/231075482-7b4ddcf4-25ee-4ae2-9c84-92ee34fdd7d9.png">

Please let us know if you have any questions.

**Comment by aspearman** [comment]
Thanks @thisisabhash!
